### PR TITLE
Object-reclaimer - new bg-worker for map cleaning in background

### LINKED ISSUE #5732: Deleting all objects in a bucket take a long time
### Environment info
- NooBaa Version: 5
- Platform: All

### Actual behavior
1. A 1TB dataset consisting of large object (~65MB) take ~9h to delete
See https://bugzilla.redhat.com/show_bug.cgi?id=1753109

### Expected behavior
1. Marked objects as deleted, don't return them in lists/counts and reclaim in bg

### PR DESCRIPTION
**Pull request by jackyalbo**
### Explain the changes
1. As part of the performance improvements. mapping deletion will not happen during the deletion of the object. The cleaning will be moved to a bg_worker which will reclaim all the deleted objects and will delete it's parts, chunks and blocks
2. Removing any object responsibility from scrubber through map_builder. Now all the object mappings cleaning will happen in object-reclaimer
3. Adding a new reclaim db flag for identifying objects that already cleaned up all of it's mappings. adding md_store support for finding those objects and for updating a large amount of objects at once - for marking as reclaimed
4. for the benefit of testing fixed md_blow to support post tiering structuring and to work in async-await instead of promises. 

### Issues: Fixed #xxx / Gap #xxx
1. Fixes #5732

### Testing Instructions:
1.

### CODE REVIEW

**Review by nimrod-becker — APPROVED** [review]

File: src/server/bg_services/objects_reclaimer.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        if (has_errors) {
+            return config.OBJECT_RECLAIMER_ERROR_DELAY;
+        }
+        return config.OBJECT_RECLAIMER_BATCH_DELAY;
+
+    }
+
+    _can_run() {
+        if (!system_store.is_finished_initial_load) {
+            dbg.log0('MirrorWriter: system_store did not finish initial load');
```
Comment (dannyzaken) [inline src/server/bg_services/objects_reclaimer.js]: first rule in school - when you copy, don't copy the name..

Comment (nimrod-becker) [reply src/server/bg_services/objects_reclaimer.js]: :sweat_smile:

**Review by dannyzaken — APPROVED** [review]

**Comment by nimrod-becker** [comment]
FYI @guymguym

File: src/tools/md_blow.js [hunk 2]
```diff
@@ -23,88 +26,98 @@ argv.key = argv.key || ('md_blow-' + Date.now().toString(36));
...
+async function blow_objects() {
     let index = 0;
 
-    function blow_next() {
+    async function blow_next() {
         if (index >= argv.count) return;
         index += 1;
-        return blow_object(index).then(blow_next);
+        await blow_object(index);
+        await blow_next();
```
Comment (guymguym) [inline src/tools/md_blow.js]: Instead of async recursion we can use a simple loop:
```javascript
while (index < argv.count) {
    index += 1;
    await blow_object(index);
}
```

File: src/test/unit_tests/coretest.js [hunk 3]
```diff
@@ -339,8 +340,9 @@ async function init_test_pools(client, system_name, pools_to_create) {
 
 // delete all test pools (including hosts and agents)
 async function clear_test_pools() {
+    console.log('CLEANING ALL CHUNKS');
+    await MDStore.instance().delete_all_chunks_in_system();
```
Comment (guymguym) [inline src/test/unit_tests/coretest.js]: I think this needs a comment that explains why we had to add it here.
It's not so easy to figure it out.

File: src/server/object_services/schemas/object_md_indexes.js [hunk 4]
```diff
@@ -115,6 +115,7 @@ module.exports = [
         // aggregate_objects_by_delete_dates()
         fields: {
             deleted: 1,
+            reclaimed: 1,
```
Comment (guymguym) [inline src/server/object_services/schemas/object_md_indexes.js]: Why did you add reclaimed to the indexed fields?
Is it needed for md-aggregator?

File: src/server/object_services/md_store.js [hunk 5]
```diff
@@ -1389,6 +1414,17 @@ class MDStore {
         });
     }
 
+    // Only for clean up in testing - Don't use unless you are sure!!!!
+    async delete_all_chunks_in_system() {
```
Comment (guymguym) [inline src/server/object_services/md_store.js]: I would add `assert(config.test_mode)` to make sure this is not called anywhere else than coretest.

File: src/server/object_services/schemas/object_md_indexes.js (hunk 4 above)
Comment (guymguym) [reply src/server/object_services/schemas/object_md_indexes.js]: Maybe you meant to add `find_unreclaimed_objects()` to the list of functions using this index?

File: src/server/object_services/md_store.js [hunk 6]
```diff
@@ -1389,6 +1414,17 @@ class MDStore {
...
+        const delete_date = new Date();
+        return this._chunks.col().updateMany({}, {
+            $set: {
+                deleted: delete_date,
+                reclaimed: delete_date
```
Comment (guymguym) [inline src/server/object_services/md_store.js]: chunks don't have a `reclaimed` property in the schema... 
What did you mean to do here?

File: src/server/object_services/md_store.js [hunk 7]
```diff
@@ -496,6 +511,16 @@ class MDStore {
         };
     }
 
+    async find_unreclaimed_objects(limit) {
+        const results = await this._objects.col().find({
```
Comment (guymguym) [inline src/server/object_services/md_store.js]: On second thought, we should create a separate index for this.
This should be a simple index with `fields: { deleted: 1 }` and `partialFilterExpression: { deleted: { $exists: true }, reclaimed: null }` and we probably want to make sure to give a hint to use it here.

Comment (guymguym) [reply src/server/object_services/md_store.js]: The reason is that this "deleted but not yet reclaimed" list should usually be very small.
So there is no real overhead of this index size, and only deletions will insert to the index.

File: src/server/object_services/md_store.js [hunk 8]
```diff
@@ -1058,15 +1083,15 @@ class MDStore {
...
      * @returns {Promise<nb.ID[]>}
      */
-    async find_parts_chunk_ids(obj, none_deleted_only) {
+    async find_parts_chunk_ids(obj) {
         const find = {
             obj: obj._id,
+            deleted: null,
         };
-        if (none_deleted_only) find.deleted = null;
         return this._parts.col().find(find, {
```
Comment (guymguym) [inline src/server/object_services/md_store.js]: Can you add a hint to this query to make sure it uses the index we intended it to use?

**Review by guymguym — COMMENTED** [review]
Reviewed and added few comments
